src/Integration/LocalServices/ErrorListInfoBarController.cs

```diff
@@ -462,6 +498,194 @@ private ProjectViewModel FindProject(Uri serverUri, string projectKey)
+                }, token);
+            }
+
+            private bool IsUpdateRequired(BoundSonarQubeProject binding, IEnumerable<LanguageGroup> projectLanguageGroups, CancellationToken token)
+            {
+                Debug.Assert(binding != null);
+
+                ConnectionInformation connection = binding.CreateConnectionInformation();
+                Dictionary<LanguageGroup, QualityProfile> newProfiles;
+                if (!this.TryGetNewProfiles(binding, projectLanguageGroups, token, connection, out newProfiles))
+                {
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckFailed);
+                    return false; // Error, can't proceed
+                }
+
+                if (!newProfiles.Keys.All(binding.Profiles.ContainsKey))
+                {
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckSolutionRequiresMoreProfiles);
+                    return true; // Missing profile, refresh
+                }
+
+                foreach (var keyValue in binding.Profiles)
+                {
+                    LanguageGroup group = keyValue.Key;
+                    ApplicableQualityProfile oldProfileInfo = keyValue.Value;
+
+                    if (!newProfiles.ContainsKey(group))
+                    {
+                        // Not a relevant profile, we should just ignore it.
+                        continue;
+                    }
+
+                    QualityProfile newProfileInfo = newProfiles[group];
+                    if (this.HasProfileChanged(newProfileInfo, oldProfileInfo))
+                    {
+                        return true;
+                    }
+                }
+
+                VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckQualityProfileIsUpToDate);
+                return false; // Up-to-date
+            }
+
+            private bool HasProfileChanged(QualityProfile newProfileInfo, ApplicableQualityProfile oldProfileInfo)
+            {
+                if (!QualityProfile.KeyComparer.Equals(oldProfileInfo.ProfileKey, newProfileInfo.Key))
+                {
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckDifferentProfile);
+                    return true; // The profile change to a different one
+                }
+
+                if (oldProfileInfo.ProfileTimestamp != newProfileInfo.QualityProfileTimestamp)
+                {
+                    VsShellUtils.WriteToGeneralOutputPane(this.host, Strings.SonarLintProfileCheckProfileUpdated);
+                    return true; // The profile was updated
+                }
+
+                return false;
+            }
+
+            private bool TryGetNewProfiles(BoundSonarQubeProject binding, IEnumerable<LanguageGroup> projectLanguageGroups, CancellationToken token, ConnectionInformation connection, out Dictionary<LanguageGroup, QualityProfile> newProfiles)
```

Comment:
"NewProfiles" is implies it is only get the profiles that are new which isn't the case; TryGetLatestProfiles/TryGetServerProfiles would be clearer.